tests/pytest_tests/unit_tests/test_launch/test_runner/test_abstract_runner.py

```diff
@@ -0,0 +1,42 @@
+import os
+from unittest.mock import MagicMock
+
+from wandb.sdk.launch.runner.abstract import AbstractRunner
+
+
+class TestRunner(AbstractRunner):
+    """Runner used to test behavior of the base class."""
+
+    def __init__(self) -> None:
+        """Initialize the test runner."""
+        super().__init__(MagicMock(), MagicMock())
+
+    def run(self, *args, **kwargs) -> None:
+        """Dummy run method for testing."""
+        pass
+
+
+def test_macro_sub():
+    """Test that macros are substituted correctly."""
+    project = MagicMock()
+    project.resource_args = {
+        "image": "${wandb_image}",
+        "memory": "1GB",
+        "gpu": "${MY_VAR}",
+        "env": {"FOO": "${wandb_project}"},
+    }
```

Comment:
If I'm reading the change correctly it looks like it should support cases where the variable is only part of the string. We might want to test that case, i.e. `"env": {"FOO": "some_text_${wandb_project}_more_text"}` -> `some_text_test-project_more_text`

Reply:
Yeah that is the intent, and you are definitely right I will add more test cases